lib/strtcpy.h

```diff
@@ -44,7 +44,7 @@
  */
 
 
-#define STRTCPY(dst, src)  strtcpy(dst, src, SIZEOF_ARRAY(dst))
+#define STRTCPY(dst, src)  strtcpy(dst, src, NITEMS(dst))
```

Comment:
But this should only ever be used on arrays of byte-sized elements, right?

strtcpy won't account for the sizeof(*dst).

Reply:
Yes and no.

STRTCPY() is designed to be used for copying both single-byte strings and multi-byte strings (UTF8).  So, there are multi-byte entities.  But yeah, they're both stored in arrays of `char`, so in the sense of the C language, these handle bytes.

But it would be conceivable that one may copy this source code to develop a WCSTCPY() macro (and the corresponding wcstcpy() function) to copy wide strings with truncation.  I prefer to give good example in our implementation, so that it'll be straight-forward to implement them as:


```c
#define WCSTCPY(dst, src)  wcstcpy(dst, src, NITEMS(dst))

inline ssize_t
wcstcpy(wchar_t *restrict dst, const wchar_t *restrict src, size_t dsize)
{
	bool     trunc;
	size_t   dlen, slen;
	wchar_t  *p;

	if (dsize == 0)
		return -1;

	slen = wcsnlen(src, dsize);
	trunc = (slen == dsize);
	dlen = slen - trunc;

	p = wmempcpy(dst, src, dlen);
	*p = '\0';

	return trunc ? -1 : slen;
}
```

Symmetry is a good thing.  :)

Reply:
So, yes, in practice, both SIZEOF_ARRAY() and NITEMS() will give the same result.  But one is more appropriate than the other.

Reply:
Oh, and a curious --and a bit unrelated-- fact I learnt a few days ago:

swprintf(3) was designed correctly, unlike snprintf(3).  It shares API with this wcstcpy().  It is not limited by an `int`, but rather `size_t`; and returns -1 on truncation instead of the length of the would-be string.

Reply:
FWIW I do not consider "length of the total string it would have tried to create" to be a bad return value.  In fact I quite like it.  But you do have to know to look for it.

Reply:
It has several problems:

-  Requires that snprintf(3) reads until the end, which makes it vulnerable to DoS attacks, very much like strlcpy(3).  Okay, one should probably validate an untrusted string before passing it to snprintf(3); but then it encourages programmers to design the functions that should be used for validating such strings with a compatible API by symmetry, as if it were a good thing to return (does it sound like OpenBSD's strlcpy(3)? :D)  (and symmetry strikes again! :).

   -  swprintf(3) can stop reading when it knows it has truncated.  Anyway, it'll return -1.

-  It is prone to off-by-one bugs.  One may accidentally write `if (snprintf(dst, ...) > sizeof(dst))` instead of using `>=`.  Not only off-by-ones; it can also happen an off-by-many, using accidentally `==`.  You may say it's so interiorized that it doesn't really happen.  I've seen this bug happen a couple of weeks ago by a libc maintainer (in strlcpy(3), not snprintf(3), but it's the same bug): <https://lore.kernel.org/linux-man/ZU4SDh-Se5gjPny5@debian/T/#m7c213beac8ddc019ec2fa0922a0cd500eb6dc65a>.  See the end of that email.

   -  swprintf() is instead checked as `if (swprintf(dst, ...) == -1)`, which is hard to get wrong.

In case it doesn't fail, it still returns the length, so in useful cases you still have it.